crates/apollo-compiler/src/validation/selection.rs

```diff
@@ -29,15 +31,142 @@ pub fn validate_selection(
     diagnostics
 }
 
+fn validate_selection_field(
+    db: &dyn ValidationDatabase,
+    field: Arc<hir::Field>,
+    diagnostics: &mut Vec<ApolloDiagnostic>,
+) {
+    let leaf = field.selection_set.selection.is_empty();
+    let mut leaf_validation_error: Option<&str> = None;
+    let field_type = field.ty(db.upcast());
+    if let Some(field_type) = field_type {
+        let type_name = field_type.name();
+        let type_def = db.find_type_definition_by_name(type_name.clone());
+        if let Some(type_def) = type_def {
+            match type_def {
+                TypeDefinition::EnumTypeDefinition(_) if !leaf => {
+                    leaf_validation_error =
+                        Some("This field is an enum and cannot select any fields");
+                }
+                TypeDefinition::ScalarTypeDefinition(_) if !leaf => {
+                    leaf_validation_error =
+                        Some("This field is a scalar and cannot select any fields");
+                }
+                TypeDefinition::ObjectTypeDefinition(_) if leaf => {
+                    leaf_validation_error = Some("This field is an object and must select fields");
+                }
+                TypeDefinition::InterfaceTypeDefinition(_) if leaf => {
+                    leaf_validation_error =
+                        Some("This field is an interface and must select fields");
+                }
+                TypeDefinition::UnionTypeDefinition(_) if leaf => {
+                    leaf_validation_error = Some("This field is an union and must select fields");
+                }
+                _ => {}
+            };
+            if let Some(error) = leaf_validation_error {
+                let name = field.name.src.clone();
+                let diagnostic_data = if leaf {
+                    DiagnosticData::MandatorySubselection {
+                        name,
+                        ty: type_name.clone(),
+                    }
+                } else {
+                    DiagnosticData::NoSubselectionAllowed {
+                        name,
+                        ty: type_name.clone(),
+                    }
+                };
+                let diagnostic = ApolloDiagnostic::new(db, field.loc.into(), diagnostic_data)
+                    .label(Label::new(field.loc, error));
+                let diagnostic = if let Some(type_def_loc) = type_def.loc() {
+                    diagnostic.label(Label::new(
+                        type_def_loc,
+                        format!("`{}` declared here", &type_name),
+                    ))
+                } else {
+                    diagnostic
+                };
+                diagnostics.push(diagnostic);
+            }
+        }
+    }
+
+    if leaf_validation_error.is_none() {
```

Comment:
If the leaf validation fails, I did not go into the validation of the field itself as it was much noise.
For example, the `SelectionOfEnum` query is the test would have validation errors: one about the sub selection, and one about the unknown `name` field.

Reply:
I think it makes sense to integrate this function into `validate_field`, as leaf field validation is part of field validation in the spec. Additionally we make use of the existing derived query like that. The leaf field validation could be integrated here (L78) (or merge that L20 alternatively since L78 also just requires the field to exist on the type, but with a different call):
https://github.com/apollographql/apollo-rs/blob/49043027fca3e1c16dbf3437832a629afbd2e00b/crates/apollo-compiler/src/validation/field.rs#L78-L85

This way you can avoid going into `validate_selection_set` for the selection set of an enum in case it is supposed to be a leaf field, also reducing the noise but keeping field validation in the same place, also you have all the data like `field_ty` and `field_definition` (L20) there already.

Just an idea, what's your opinion on that? 🙂

Reply:
Yes, that's a good idea.
Please check https://github.com/apollographql/apollo-rs/pull/465/commits/5ee19ce54212a45668c1d8ab48cee86a76dcfce8
I could not re-use `field_definition` as I need the type definition.